README
Crypt-OpenSSL-Blowfish version 0.02
===================================

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the
README file from a module distribution so that people browsing the
archive can use it get an idea of the modules uses. It is usually a
good idea to provide version information here so that people can
decide whether fixes for the module are worth downloading.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

SYNOPSIS

   use Crypt::OpenSSL::Blowfish;
   my $cipher = new Crypt::Blowfish $key;
   my $ciphertext = $cipher->encrypt($plaintext);
   $plaintext = $cipher->decrypt($ciphertext);

COPYRIGHT AND LICENCE

Copyright (C) Vitaly Kramskikh

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.0 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Crypt::OpenSSL::Blowfish.

{{$NEXT}}

0.08 -- Mon Nov 27 23:07:38 AST 2023

    [Legacy Notice]
    OpenSSL version 3 considers Blowfish to be a legacy encryption
    method.  Users of this module (if any) need to consider moving to
    other encryption methods.

    [Compatibility Notice]
    This version maintains compatibility with the older versions.
    However, this is not fully compatible with openssl's bf-ecb cipher.
    See the documentation relating to how to use the compatible version.

    [Significant rewrite since 0.02]
    - OpenSSL 3 has deprecated the use of low-level functions.
    - Update to use OpenSSL v3 EVP_ api for openssl >= 3.0.0
    - Significant changes to the module while retaining compatibity with
      the previous versions.  As Blowfish is now outdated and considered
      legacy by OpenSSL this rewrite was done out of interest not necessity.

    [Detailed Change Log]
    - 5ae4ef9 Move Blowfish.pm to lib dir to properly update provides
    - aa28a27 v0.07

0.07 -- Mon Nov 27 22:39:10 AST 2023

    [Legacy Notice]
    OpenSSL version 3 considers Blowfish to be a legacy encryption
    method.  Users of this module (if any) need to consider moving to
    other encryption methods.

    [Compatibility Notice]
    This version maintains compatibility with the older versions.
    However, this is not fully compatible with openssl's bf-ecb cipher.
    See the documentation relating to how to use the compatible version.

    [Significant rewrite since 0.02]
    - OpenSSL 3 has deprecated the use of low-level functions.
    - Update to use OpenSSL v3 EVP_ api for openssl >= 3.0.0
    - Significant changes to the module while retaining compatibity with
      the previous versions.  As Blowfish is now outdated and considered
      legacy by OpenSSL this rewrite was done out of interest not necessity.

    [Detailed Change Log]
    - 7fbbaa1 Update version for release
    - dbbc13a v0.06

0.06 -- Sat Nov 25 13:07:19 AST 2023

    [Legacy Notice]
    OpenSSL version 3 considers Blowfish to be a legacy encryption
    method.  Users of this module (if any) need to consider moving to
    other encryption methods.

    [Compatibility Notice]
    This version maintains compatibility with the older versions.
    However, this is not fully compatible with openssl's bf-ecb cipher.
    See the documentation relating to how to use the compatible version.

    [Significant rewrite]
    - OpenSSL 3 has deprecated the use of low-level functions.
    - Update to use OpenSSL v3 EVP_ api for openssl >= 3.0.0
    - Significant changes to the module while retaining compatibity with
      the previous versions.  As Blowfish is now outdated and considered
      legacy by OpenSSL this rewrite was done ot of interest not necessity.

    [Detailed Change Log]
    - 538078c Update version number for a release
    - a8499df Properly free EVP_CIPHER_CTX
    - 4779460 v0.05

0.05 -- Fri Nov 24 23:28:52 AST 2023

    [Legacy Notice]
    OpenSSL version 3 considers Blowfish to be a legacy encryption
    method.  Users of this module (if any) need to consider moving to
    other encryption methods.

    [Compatibility Notice]
    This version maintains compatibility with the older versions.
    However, this is not fully compatible with openssl's bf-ecb cipher.
    See the documentation relating to how to use the compatible version.

    [Significant rewrite]
    - OpenSSL 3 has deprecated the use of low-level functions.
    - Update to use OpenSSL v3 EVP_ api for openssl >= 3.0.0
    - Significant changes to the module while retaining compatibity with
      the previous versions.  As Blowfish is now outdated and considered
      legacy by OpenSSL this rewrite was done ot of interest not necessity.

    [Detailed Change Log]
    - df6fa5c Update version number for a release
    - 33ef340 Fix issues with the conversion of endianness
    - f766537 v0.04

0.04 -- Thu Nov 23 20:19:45 AST 2023

    [Legacy Notice]
    OpenSSL version 3 considers Blowfish to be a legacy encryption
    method.  Users of this module (if any) need to consider moving to
    other encryption methods.

    [Compatibility Notice]
    This version maintains compatibility with the older versions.
    However, this is not fully compatible with openssl's bf-ecb cipher.
    See the documentation relating to how to use the compatible version.

    [Significant rewrite]
    - OpenSSL 3 has deprecated the use of low-level functions.
    - Update to use OpenSSL v3 EVP_ api for openssl >= 3.0.0
    - Significant changes to the module while retaining compatibity with
      the previous versions.  As Blowfish is now outdated and considered
      legacy by OpenSSL this rewrite was done ot of interest not necessity.

    [Detailed Change Log]
    - 4c6d166 Update .gitignore
    - ac0eba7 Update version number for a release
    - 271ca04 Add diagonistics to tests
    - 212b609 Don't include hexdump by default
    - 9d91b25 Also search for LibreSSL
    - ead3ad5 v0.03
    - 8e5ea58 Update Changes for a release

0.03 -- Wed Nov 22 23:48:06 AST 2023

    [Legacy Notice]
    OpenSSL version 3 considers Blowfish to be a legacy encryption
    method.  Users of this module (if any) need to consider moving to
    other encryption methods.

    [Compatibility Notice]
    This version maintains compatibility with the older versions.
    However, this is not fully compatible with openssl's bf-ecb cipher.
    See the documentation relating to how to use the compatible version.

    [Significant rewrite]
    - OpenSSL 3 has deprecated the use of low-level functions.
    - Update to use OpenSSL v3 EVP_ api for openssl >= 3.0.0
    - Significant changes to the module while retaining compatibity with
      the previous versions.  As Blowfish is now outdated and considered
      legacy by OpenSSL this rewrite was done ot of interest not necessity.

    [Detailed Change Log]
    - c9dd076 Add .gitignore file

Blowfish.pm
package Crypt::OpenSSL::Blowfish;

use strict;
use Carp;

use vars qw/$VERSION @ISA/;

require DynaLoader;
@ISA = qw/DynaLoader/;

$VERSION = '0.07';

bootstrap Crypt::OpenSSL::Blowfish $VERSION;

sub blocksize   {  8; }
sub keysize     {  0; }
sub min_keysize {  8; }
sub max_keysize { 56; }

sub encrypt {
    my ($self, $data) = @_;

    return $self->crypt($data, 1);

}

sub decrypt {
    my ($self, $data) = @_;

    return $self->crypt($data, 0);

}

1;

__END__

=head1 NAME

Crypt::OpenSSL::Blowfish - Blowfish Algorithm using OpenSSL

=head1 SYNOPSIS

    use Crypt::OpenSSL::Blowfish;

    my $cipher = Crypt::OpenSSL::Blowfish->($key, {});

    or (for compatibility.with 0.02 and below)

    my $cipher = Crypt::OpenSSL::Blowfish->($key); # DON'T do this

    my $ciphertext = $cipher->encrypt($plaintext);
    my $plaintext = $cipher->decrypt($ciphertext);

=head1 BLOWFISH IS CONSIDERED LEGACY BY OPENSSL v3

This module has been updated with support for OpenSSL v3 but should
be considered to be end of life.  OpenSSL requires the I<legacy> provider
to be specifically loaded to use Blowfish.  It is a deprecated encryption
module and you B<need> to move to a different encryption algorithm.

=head1 COMPATIBILITY WARNING

B<WARNING> Version 0.02 and below B<DO NOT> produce OpenSSL compatible
encryption or decrypt OpenSSL encrypted messages properly.

Basically, those versions call BF_encrypt and BF_decrypt without properly
converting the input to big endian and the output to little endian where
needed.

Version 0.03 and above correctly calls BF_ecb_encrypt and BF_ecb_decrypt
or the EVP_* functions but I<DEFAULTS> to the B<incorrect> method for
compatibility.  This B<MAY> change in subsequent versions.

To obtain the correct method you B<must> provide a second option in the new
constructor.  Even an empty hash will work and will be compatible with later
versions.

    my $cipher = Crypt::OpenSSL::Blowfish($key, {});

The encryption/decryption method is then compatible with the BF_ecb_encrypt
and BF_ecb_decrypt and well as the new EVP_ methods using B<EVP_bf_ecb> cipher
when compiled with openssl v3.

=head1 DESCRIPTION

Crypt::OpenSSL::Blowfish implements the Blowfish Algorithm using functions
contained in the OpenSSL crypto library.  Crypt::OpenSSL::Blowfish has an
interface similar to Crypt::Blowfish, but produces different result than
Crypt::Blowfish. This is no longer correct if you use the new method to
instantiate Crypt::OpenSSL::Blowfish.  Using the new method results in
encryption which is compatible with Crypt::Blowfish.

=head1 METHODS

=head2 new($key, ... )

    # New method - openssl compatible
    my $cipher = Crypt::OpenSSL::Blowfish->new($key, {});

    # Old method - NOT openssl compatible
    my $cipher = Crypt::OpenSSL::Blowfish->new($key);

=head2 encrypt(data)

    my $encrypted = $cipher->encrypt($plaintext);

=head2 encrypt(data)

    my $decrypted = $cipher->decrypt($ciphertext);

=head2 get_little_endian(data)

Converts the data to little-endian format. Required to convert the old
encryption to openssl compatible bf-ecb/EVP_bf_ecb.

See t/upgrade.t for examples to handle the endian conversion and encryption
upgrades.

    my $leval = get_little_endian(data)

=head2 get_big_endian(data)

Converts the data to little-endian format. Required to convert the old
encryption to openssl compatible bf-ecb/EVP_bf_ecb

See t/upgrade.t for examples to handle the endian conversion and encryption
upgrades.

    my $beval = get_big_endian(data)

=head1 SEE ALSO

L<Crypt::Blowfish>

http://www.openssl.org/

=head1 AUTHOR

Vitaly Kramskikh, E<lt>vkramskih@cpan.orgE<gt>
Timothy Legge, E<lt>timlegge@cpan.orgE<gt>

=head1 LICENSE

No license was mentioned in the original version 0.01 or 0.02. Based
on the changes made by TIMLEGGE, 0.03 and forward will be as follows:

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.5 or,
at your option, any later version of Perl 5 you may have available.

In the unlikely chance that that is an issue for anyone feel free to
contact TIMLEGGE

=cut


Makefile.PL
# This Makefile.PL for Crypt-OpenSSL-Blowfish was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.49.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.008;
use ExtUtils::MakeMaker;

use Config;
use File::Spec;
use Crypt::OpenSSL::Guess qw(openssl_version openssl_inc_paths openssl_lib_paths);

my %args;

my ($major, $minor, $patch) = openssl_version();
print "Installed OpenSSL: $major.$minor", defined $patch ? $patch : "", "\n";

$args{INC} =  openssl_inc_paths();
$args{LIBS} = [openssl_lib_paths() . ' -lssl -lcrypto'];

my $cc_option_flags;

if ($Config::Config{cc} =~ /gcc/i) {
  $cc_option_flags .= $ENV{AUTHOR_TESTING} ? ' -Wall -Wpointer-sign' : ' -Wall';
} else {
  $cc_option_flags .= $ENV{AUTHOR_TESTING} ? ' -Wall -Wpointer-sign' : '';
}

if ($Config{gccversion} =~ /llvm/i) {
  if ( $^O eq 'darwin' && (
    $Config{gccversion} =~ /LLVM 1[2-9].\d.[5-9]/ ||
    $Config{gccversion} =~ /LLVM 1[3-9]/ ))
  {
    $cc_option_flags .= ' -Wno-deprecated-declarations -Wno-compound-token-split-by-macro';
  } else {
    $cc_option_flags .= ' -Wno-deprecated-declarations';
  }

  if ($Config{gccversion} =~ /llvm 13/i) {
      $cc_option_flags .= ' -Wno-compound-token-split-by-macro';
  }

  if ($Config{PERL_VERSION} <= 20) {
    $cc_option_flags .= ' -Wno-error=pointer-bool-conversion';
  }
}

if ($Config{myuname} =~ /sunos|solaris/i) {
  $args{OPTIMIZE} = $cc_option_flags;
} elsif ($Config{myuname} =~ /strawberry-{0,1}perl/i) {
  if ($Config{PERL_VERSION} <= 24) {
    $args{LIBS} = [openssl_lib_paths() . ' -leay32'];
  }
} elsif ($Config{myuname} =~ /darwin/ ) {
  $args{LDDLFLAGS} = openssl_lib_paths() . " " . $Config{lddlflags};
} else {
  $args{OPTIMIZE} = $cc_option_flags;
}

my %WriteMakefileArgs = (
  "ABSTRACT" => "Blowfish Algorithm using OpenSSL",
  "AUTHOR" => "Vitaly Kramskikh <vkramskih\@cpan.org>, Timothy Legge <timlegge\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Crypt::OpenSSL::Guess" => "0.10"
  },
  "CONFIGURE_REQUIRES" => {
    "Crypt::OpenSSL::Guess" => "0.10",
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Crypt-OpenSSL-Blowfish",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Crypt::OpenSSL::Blowfish",
  "PREREQ_PM" => {},
  "TEST_REQUIRES" => {
    "MIME::Base64" => 0,
    "Test2::V0" => 0
  },
  "VERSION" => "0.07",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

%WriteMakefileArgs = (
    %WriteMakefileArgs,
    %args,
);

my %FallbackPrereqs = (
  "Crypt::OpenSSL::Guess" => "0.10",
  "MIME::Base64" => 0,
  "Test2::V0" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
